### DISCUSSION TITLE
Utility function to get the SQL type from psycopg.Column object

### DISCUSSION THREAD

**Discussion by johann-petrak** [psycopg/psycopg#449] [Q&A]
See https://www.psycopg.org/psycopg3/docs/api/objects.html#psycopg.Column

This represents the SQL column using a numerical type number and other fields which are relevant for some types and empty for others which is clumsy to use.

Is there an utility method somewhere to convert such an object into the actual SQL DDL representation, e.g. "field1 VARCHAR(10) NOT NULL" or "field2 DOUBLE PRECISION" ?

Also the field `null_ok` is None, though each column should either be defined as NULL or NOT NULL, so why is it not either True or False?

**Comment by dvarrazzo** [comment]
Hello,

`null_ok` is not an information we receive from the libpq as the result of a query. So I'm afraid we cannot return anything useful there.

The numerical type number can be decoded using the `adapters.types` map on the cursor:

```python
>>> cur = conn.execute("select null::float8, null::varchar(10)")
>>> cur.description[0].type_code
>>> cur.adapters.types[701]
<TypeInfo: float8 (oid: 701, array oid: 1022)>
>>> cur.adapters.types[701].name
'float8'
>>> cur.adapters.types[cur.description[1].type_code].name
'varchar'
```

The problem is the modifier. Unfortunately this information depends on the type. Psycopg has a bit of knowledge, so actually if you see the `repr()` of the columns you will see:

```python
>>> cur.description
[<Column 'float8', type: float8 (oid: 701)>,
 <Column 'varchar', type: varchar(10) (oid: 1043)>]
```

but [it's a bit hacked up](https://github.com/psycopg/psycopg/blob/d2e6a3b73b1b086b1e03554f0d9b0e181e2457b2/psycopg/psycopg/_column.py#L62-L69) so it's only used for repr display and there's no official interface way to take that info out yet.

There are several builtin fields which accept a modifier: this can be seen in the system catalog:

```
piro=# select typname, typmodin, typmodout from pg_type where (typmodin::text <> '-' or typmodout::text <> '-') and not typname ~ '^_';
┌─────────────┬─────────────────────┬──────────────────────┐
│   typname   │      typmodin       │      typmodout       │
├─────────────┼─────────────────────┼──────────────────────┤
│ bpchar      │ bpchartypmodin      │ bpchartypmodout      │
│ varchar     │ varchartypmodin     │ varchartypmodout     │
│ time        │ timetypmodin        │ timetypmodout        │
│ timestamp   │ timestamptypmodin   │ timestamptypmodout   │
│ timestamptz │ timestamptztypmodin │ timestamptztypmodout │
│ interval    │ intervaltypmodin    │ intervaltypmodout    │
│ timetz      │ timetztypmodin      │ timetztypmodout      │
│ bit         │ bittypmodin         │ bittypmodout         │
│ varbit      │ varbittypmodin      │ varbittypmodout      │
│ numeric     │ numerictypmodin     │ numerictypmodout     │
└─────────────┴─────────────────────┴──────────────────────┘
```

so for psycopg it's an open problem: we can learn this handful from the system catalog (probably psycopg already works well with them, I can't vouch for the most exotic types but the common ones should be alright), however it's not impossible for an extension type to have its modifier function and it would be hard for psycopg to learn it.

using `psql -E`, it can be seen that psql decodes that field using the utility function `pg_catalog.format_type(a.atttypid, a.atttypmod)` (where a is the pg_attribute entry of the field). These two field details are actually returned by the result: by `PGresult.ftype()` (exposed as `Column.type_code`) and `PGresult.fmod()` (not exposed by the column), so the most robust way to obtain that info is to query the server again:

```
cur2 = conn.cursor()
cur2.executemany(
    "select pg_catalog.format_type(%s, %s)",
    [(col.type_code, cur.pgresult.fmod(i)) for i, col in enumerate(cur.description)],
    returning=True)

while 1:
    print(cur2.fetchone()[0])
    if not cur2.nextset():
        break

double precision
character varying(10)
```

---

To summarize:

1) one important design choice of psycopg is to be explicit when it does extra query. So, unfortunately we don't want to introduce utility functions calling the server, for e.g. `cur.description[1].get_pretty_type()`, because that could mess up the state of the transaction, can fail in many different way, has to take account of async connections... We don't want to do that.

2) as a consequence, any info we provide, can be based only on the result by a query. We have the type oid and the type modifier. We could try to decode them better (by adding the proper decode method to the [`TypeInfo` object](https://github.com/psycopg/psycopg/blob/d2e6a3b73b1b086b1e03554f0d9b0e181e2457b2/psycopg/psycopg/_typeinfo.py#L27), instead of hacking up the Column object more), but only for builtin types (which, tbh, would cover the 99.9% of the cases).

3) as you are writing a program to know these info, you can definitely run more queries. Information such as whether a column is nullable or calling type modifier decoding function is therefor available for you. Pretty much everything is in the [pg_attribute system table](https://www.postgresql.org/docs/current/catalog-pg-attribute.html) (for instance `attnotnull` answers your nullable question).

4) you can run `psql -E` and type `\d table_name`: this will print the queries that psql runs in order to display its result and will allow you to reverse-engineer the catalog a bit.

Hope this helps.

**Reply by johann-petrak** [reply]
Thank you so much, this is very helpful!
